controller/contour/job.go

```diff
@@ -0,0 +1,197 @@
+		if errors.IsNotFound(err) {
+			return false, nil, nil
+		}
+		return false, nil, err
+	}
+	return true, current, nil
+}
+
+// desiredJob generates the desired Job for the given contour.
+func desiredJob(contour *operatorv1alpha1.Contour) (*batchv1.Job, error) {
+	one := int32(1)
+	zero := int32(0)
+
+	job := &batchv1.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      contour.Name + "-certgen",
+			Namespace: contour.Spec.Namespace.Name,
+		},
+		Spec: batchv1.JobSpec{
+			Parallelism:  &one,
+			Completions:  &one,
+			BackoffLimit: &one,
+			// Make job eligible to for immediate deletion.
+			TTLSecondsAfterFinished: &zero,
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{"app": jobLabelValue},
+				},
+				Spec: corev1.PodSpec{
+					RestartPolicy: corev1.RestartPolicyNever,
+				},
+			},
+		},
+	}
+
+	// TODO [danehans]: Once https://github.com/projectcontour/contour-operator/pull/30/files
+	// merges, associate contour controller owning label.
+	/*job.Labels = map[string]string{
+		// associate the job with the contour
+		controllerJobLabel: contour.Name,
+	}*/
+
+	// TODO [danehans]: Pass in svcAct name to func when https://github.com/projectcontour/contour-operator/pull/21 merges.
+	job.Spec.Template.Spec.ServiceAccountName = jobSvcAcctName
+
+	user := int64(65534)
+	group := int64(65534)
+	nonRoot := true
+	secCtx := corev1.PodSecurityContext{
+		RunAsUser:    &user,
+		RunAsGroup:   &group,
+		RunAsNonRoot: &nonRoot,
+	}
+	job.Spec.Template.Spec.SecurityContext = &secCtx
+
+	nsEnv := corev1.EnvVar{
+		Name: jobNsEnvVar,
+		ValueFrom: &corev1.EnvVarSource{
+			FieldRef: &corev1.ObjectFieldSelector{
+				FieldPath: "metadata.namespace",
+			},
+		},
+	}
+
+	ctrContainer := corev1.Container{
+		Name:            jobContainerName,
+		Image:           jobImageName,
+		ImagePullPolicy: corev1.PullAlways,
+		Command: []string{
+			"contour",
+			"certgen",
+			"--kube",
+			"--incluster",
+			"--overwrite",
+			"--secrets-format=compact",
+			fmt.Sprintf("--namespace=$(%s)", jobNsEnvVar),
+		},
+		Env: []corev1.EnvVar{nsEnv},
+	}
+
+	job.Spec.Template.Spec.Containers = []corev1.Container{ctrContainer}
+
+	// TODO [danehans]: Once https://github.com/projectcontour/contour-operator/pull/30/files
+	// merges, associate contour controller owning label.
+	//job.SetOwnerReferences([]metav1.OwnerReference{newContourOwnerRef(contour)})
+
+	return job, nil
+}
+
+// updateJob updates a Job based if current does not match desired.
+func (r *Reconciler) updateJob(ctx context.Context, current, desired *batchv1.Job) error {
+	changed, updated := jobConfigChanged(current, desired)
+	if !changed {
+		return nil
+	}
+
+	// Diff before updating because the client may mutate the object.
+	diff := cmp.Diff(current, updated, cmpopts.EquateEmpty())
+	if err := r.Client.Update(ctx, updated); err != nil {
+		return fmt.Errorf("failed to update job %s/%s: %w", updated.Namespace, updated.Name, err)
+	}
```

Comment:
Most job fields are read-only. We probably need to just delete the job here.

Reply:
@jpeach the [k8s docs](https://kubernetes.io/docs/concepts/workloads/controllers/job/#job-termination-and-cleanup) state:
> When a Job completes, no more Pods are created, but the Pods are not deleted either. Keeping them around allows you to still view the logs of completed pods to check for errors, warnings, or other diagnostic output.

It seems like keeping the job around does provide value. Should the operator simply ignore any changes to the job?

Reply:
We hit this problem with the example deployment. We used to just ship YAML with the job spec and when you reapplied it (with the updated image field), changing the job spec would fail because job fields are read-only.

You can keep the job around if you like, but you can't update it. I see that you changed to `recreateJob`, which looks like the right approach.